pennylane/math/is_independent.py

```diff
@@ -0,0 +1,358 @@
+
+
+def _is_indep_numerical(func, interface, args, kwargs, num_pos, seed, atol, rtol, bounds):
+    """Test whether a function returns the same output at random positions.
+
+    Args:
+        func (callable): Function to be tested
+        interface (str): Interface used by ``func``
+        args (tuple): Positional arguments with respect to which to test
+        kwargs (dict): Keyword arguments for ``func`` at which to test;
+            The ``kwargs`` are kept fix in this test.
+        num_pos (int): Number of random positions to test
+        seed (int): Seed for random number generator
+        atol (float): Absolute precision for comparing the outputs
+        rtol (float): Absolute precision for comparing the outputs
+        bounds (tuple[int, int]): Limits of the range from which to sample
+
+    Returns:
+        bool: Whether ``func`` returns the same output at the randomly
+        chosen points.
+    """
+    rnd_args = _get_random_args(args, interface, num_pos, seed, bounds)
+    original_output = func(*args, **kwargs)
+    is_tuple_valued = isinstance(original_output, tuple)
+    for _rnd_args in rnd_args:
+        new_output = func(*_rnd_args, **kwargs)
+        if is_tuple_valued:
+            if not all(
+                np.allclose(new, orig, atol=atol, rtol=rtol)
+                for new, orig in zip(new_output, original_output)
+            ):
+                return False
+        else:
+            if not np.allclose(new_output, original_output, atol=atol, rtol=rtol):
+                return False
+
+    return True
+
+
+def is_independent(
+    func,
+    interface,
+    args,
+    kwargs=None,
+    num_pos=5,
+    seed=9123,
+    atol=1e-6,
+    rtol=0,
+    bounds=(-np.pi, np.pi),
+):
+    """Test whether a function is independent of its input arguments,
+    both numerically and analytically.
+
+    Args:
+        func (callable): Function to be tested
+        interface (str): Interface used by ``func`` and the tests
+        args (tuple): Positional arguments with respect to which to test
+        kwargs (dict): Keyword arguments for ``func`` at which to test;
+            The ``kwargs`` are kept fixed in this test.
+        num_pos (int): Number of random positions to test
+        seed (int): Seed for random number generator
+        atol (float): Absolute precision for comparing the outputs
+        rtol (float): Absolute precision for comparing the outputs
+        bounds (tuple[int, int]): Limits of the range from which to sample
+
+    Returns:
+        bool: Whether ``func`` returns the same output at randomly
+        chosen points and is numerically independent of its arguments.
+
+    .. warning::
+
+        This function is experimental.
+        As such, it might yield wrong results and might behave
+        slightly differently in distinct ``interface``s for some edge cases.
+        A currently known edge case is the function
+        ``lambda x: x if abs(x) <1e-5 else 0. * x`` at ``x=0.0``.
```

Comment:
```suggestion
        For example, a currently known edge case are piecewise
        functions that use classical control, such as 
        
        .. code-block:: python
        
            def func(x):
                if abs(x) <1e-5:
                    return x
                else:
                    return 0. * x.
```

- Using `.. code-block::` means we can have syntax highlighting :)
- Explicit function definition is nice, since a reader will realize this corresponds to the input `func`.
- Might be good to also very briefly explain the pattern to avoid (feel free to edit the intro sentence I have!)

Reply:
Ah, good call!